Publish Safe Module exploit analysis (#87)
commonprefix/website: blog/4m-usd-squid-router-module-safe-exploit-analysis.md

---
title: "$4M SquidRouterModule Exploit: What Exactly Happened"
date: "26/05/2026"
desc: "Analysis of a third-party Safe module exploit on May 25, 2026."
authors: "nikolas,ristic"
index: true
---

**All Axelar user funds, approvals, and integrations are unaffected, and no action is needed. Our integration partner Squid was also not affected.**

## TL;DR

> - In May 2026, an exploit targeted a third-party Safe module deployed at the same address on [Ethereum](https://etherscan.io/address/0x1f1d37a3Bf840e35c6a860c7C2dA71Fe555123ca), [Base](https://basescan.org/address/0x1f1d37a3Bf840e35c6a860c7C2dA71Fe555123ca), and [Arbitrum](https://arbiscan.io/address/0x1f1d37a3Bf840e35c6a860c7C2dA71Fe555123ca): `0x1f1d37a3Bf840e35c6a860c7C2dA71Fe555123ca`.

Reviewer: ```suggestion
- On May 25, 2026, an exploit targeted a third-party [Safe](https://safe.global/) module deployed at the same address on [Ethereum](https://etherscan.io/address/0x1f1d37a3Bf840e35c6a860c7C2dA71Fe555123ca), [Base](https://basescan.org/address/0x1f1d37a3Bf840e35c6a860c7C2dA71Fe555123ca), and [Arbitrum](https://arbiscan.io/address/0x1f1d37a3Bf840e35c6a860c7C2dA71Fe555123ca): `0x1f1d37a3Bf840e35c6a860c7C2dA71Fe555123ca`.
```

> - The observed victims were individual Safe smart accounts that had enabled the affected Safe module and granted delegate permissions through the module's permissions manager. The affected component was the affected Safe application module, not Axelar or Squid.

Reviewer: ```suggestion
- The observed victims were individual Safe smart accounts that had enabled the affected Safe module and granted delegate permissions through the module's permissions manager. Neither Axelar nor Squid were affected by this incident.
```

> - Across the indexed attack transactions, the observed drain total is approximately $3.98M across 308 attack transactions, and 88 distinct victim Safes, using the snapshot from 26th of May 2026.

Reviewer: ```suggestion
- Across the attack transactions we indexed on May 26th 2026, the observed drain total is approximately $3.98M across 308 attack transactions and 88 distinct victim Safes.
```

- The core issue was that the affected Safe module decoded a `delegate` address from attacker-controlled cross-chain payload bytes and then used that address for local Safe permission checks.

> - After the drains, a large portion of the proceeds was consolidated across Ethereum, Base, and Arbitrum via Relay and sent as 3,070,767.09 DAI to attacker wallet 5.

Reviewer: ```suggestion
- Following the series of exploit transactions, attack proceeds were consolidated across Ethereum, Base, and Arbitrum via Relay and sent as 3,070,767.09 DAI to `[0xa447f71782135ab96a71374271a749ff7aa54859](https://etherscan.io/address/0xa447f71782135ab96a71374271a749ff7aa54859) on Ethereum.
```

## Incident Summary

> On May 25, 2026, 88 distinct victim Safes were observed drained, totaling approximately $3.98M worth of stolen tokens. The exploit targeted a third-party Safe module. It was written, deployed, and operated by a third-party, and was part of a larger Safe module suite whose deployer also deployed sibling modules for Aave, Morpho, Yearn, Uniswap V2/V3/V4, and Krystal.

Reviewer: ```suggestion
On May 25, 2026, 88 distinct victim Safes were drained, totaling approximately $3.98M worth of stolen tokens. The exploit targeted a third-party Safe module called `SquidRouterModule` (unaffiliated with the Squid protocol, [according to the Squid team](https://x.com/squidrouter/status/2058890710611276238?s=20)). It was written, deployed, and operated by a third-party, and was part of a larger Safe module suite whose deployer also deployed sibling modules for Aave, Morpho, Yearn, Uniswap V2/V3/V4, and Krystal.
```

> The identity of the module deployer has not been confirmed, but given a recent [Blockscan Chat message](https://etherscan.io/tx/0xa7db9793d4978a26ee01c2a6cbf5ba154925a7532e31ae731f7f08ba4295ee70) it's possible [NewMarketTrading](https://newmarkettrading.com) is the party behind these contracts.

Reviewer: ```suggestion
At the time of writing, the identity of the module deployer has not yet been confirmed, but given a recent [Blockscan Chat message](https://etherscan.io/tx/0xa7db9793d4978a26ee01c2a6cbf5ba154925a7532e31ae731f7f08ba4295ee70) it's possible [NewMarketTrading](https://newmarkettrading.com) is the deployer of `SquidRouterModule`.
```

> At the time of writing, the identity of the module deployer has not yet been confirmed, but given a recent [Blockscan Chat message](https://etherscan.io/tx/0xa7db9793d4978a26ee01c2a6cbf5ba154925a7532e31ae731f7f08ba4295ee70) it's possible [NewMarketTrading](https://newmarkettrading.com) is the deployer of `SquidRouterModule`.

Reviewer: ```suggestion
At the time of writing, the identity of the module deployer has not yet been confirmed, but given a recent [Blockscan Chat message](https://etherscan.io/tx/0xa7db9793d4978a26ee01c2a6cbf5ba154925a7532e31ae731f7f08ba4295ee70) it's possible that [NewMarketTrading](https://newmarkettrading.com) is behind the `SquidRouterModule`.
```

> The observed victims were individual Safe smart accounts. Each Safe had previously enabled the exploited module.

Reviewer: All of the safes were 1-of-1s, right? Although there's no evidence that those owner key were compromised and that even if so that would be irrelevant to this attack

Author: yes but I think it's not relevant.

Reviewer: Cool

- Attacker wallet 1: [`0x7c82cb4b2909c50c7c0f2b696eee7565e0a23bb8`](https://etherscan.io/address/0x7c82cb4b2909c50c7c0f2b696eee7565e0a23bb8), the main operator wallet. It deployed attacker contracts, executed most attacker-contract calls, used Relay during consolidation, and sent DAI to attacker wallet 5. It also deployed and used the fake `"u"` token (`0xe6ff0fe017d09d690493dec0f0f55e8f9cdc3512`) as the worthless token in its swap path.
- Attacker wallet 2: [`0x9bdc730183821b6bb2b51be30b77c964fa645b91`](https://etherscan.io/address/0x9bdc730183821b6bb2b51be30b77c964fa645b91), a co-operator wallet. It deployed attacker contracts, received two 1 ETH Tornado Cash withdrawals, and then funded attacker wallet 1 with operational gas across Ethereum, Base, and Arbitrum.
- Attacker wallet 3: [`0x7e54c729148a95bca651f3214ac9ebefd3fb1271`](https://etherscan.io/address/0x7e54c729148a95bca651f3214ac9ebefd3fb1271), which made direct calls to the affected Safe module. It was funded by `0x51dbd97fa14f4d78f8afc2e92a692a9e087efb5c` with 0.0687 ETH, and deployed 19 fake-token contracts across Ethereum, Base, and Arbitrum, including fake `"USDT"` tokens, together with matching Uniswap V3 pools.

> - Attacker wallet 4: [`0xc8ef4003d9db3863b9af26afcf2275378bfa83e4`](https://etherscan.io/address/0xc8ef4003d9db3863b9af26afcf2275378bfa83e4), a secondary operator wallet that made direct calls to the affected Safe module. It was funded through Orbiter Finance Bridge 3, including 0.9865 ETH before the exploit window and 1.9945 ETH during the operation, and used separate fake tokens, specifically `"sasx"`, `"sas"` and `"ddxx v3"`.

Reviewer: ```suggestion
- Attacker wallet 4: [`0xc8ef4003d9db3863b9af26afcf2275378bfa83e4`](https://etherscan.io/address/0xc8ef4003d9db3863b9af26afcf2275378bfa83e4), a secondary operator wallet that made direct calls to the affected Safe module. It was funded through Orbiter Finance Bridge 3, including 0.9865 ETH before the exploit window and 1.9945 ETH during the operation, and used separate fake tokens, specifically `"sasx"`, `"sas"`, and `"ddxx v3"`.
```

- Arbitrum: [`0x35b94f725bfdc1d00ef15df99dafadf6ecf3a0e9`](https://arbiscan.io/address/0x35b94f725bfdc1d00ef15df99dafadf6ecf3a0e9), deployed by attacker wallet 1, with 4 observed calls.

### Step 3. The Safe approved and swapped assets into attacker-controlled destinations

Once `SquidRouterModule` blindly accepted the payload, it called the victim Safe through `execTransactionFromModule`. The Safe then approved and swapped assets through attacker-controlled routing.

`SquidRouterModule` did not expose a generic transfer action or arbitrary Safe call. Its action menu was limited to swaps, approvals, and native wrap/unwrap operations, and approval spenders were constrained to Squid Router, supported Universal Routers, or Permit2. The attacker therefore could not simply approve an attacker wallet or transfer tokens directly out of the Safe.

The extraction path was the swap itself. By routing through attacker-created Uniswap V3 pools paired against worthless attacker-issued tokens, the Safe sent real assets into the pool and received those worthless tokens back. The attacker, as the pool LP, could then withdraw the real assets from the pool.

> The attacker used multiple fake tokens and created Uniswap V3 pools as drain destinations. One example swap on Ethereum is [`0x8de614bdb7acf5dcbdfe5ce8ed17ec2a2058e7708e6a4cf44f0e523c72df24c3`](https://etherscan.io/tx/0x8de614bdb7acf5dcbdfe5ce8ed17ec2a2058e7708e6a4cf44f0e523c72df24c3):

Reviewer: Was it not just one fake token? `u`? The other fake tokens were by other addresses which are not connected to the original attacker, no?

Reviewer: Do we mention potential copycats somewhere?

Reviewer: Which are all the fake token names used by the original attacker, if any beyond `u`?

Author: There are 3 distinct wallet groups and fake token groups. I mentioned them now. I wanted to have more info on how group 2 and 3 consolidated the funds. I have some info on how they were funded (TornadoCash + Orbiter Finance).

- ONDO: about 23,011 tokens ($9,870).
- UNI: about 2,603.6 tokens ($8,748).
- CFG: about 25,463 tokens ($7,749).
- MORPHO: about 3,009.3 tokens ($7,192).
- ENA: about 56,582 tokens ($5,718).
- FLUID: about 1,406.7 tokens ($4,220).
- USDT: about 3,533.1 tokens ($3,533).
- EUL, LBTC, ZRO, LINK, WLD, Anon, AERO, EURC, SKY, SLVon, SOL, msETH, XAUt, COOKIE, AVNT, USDT0, Cake, AUKI, and SAPIEN, together totaling about $12,171.

Base was the largest loss surface in the observed data, accounting for roughly 71% of the total. Ethereum accounted for roughly 27%, and Arbitrum accounted for roughly 2%.

Across the current dataset, 76 of the 88 observed victim Safes were drained on two or more chains.

## Fund Movement

After the drains, funds were pulled together across Ethereum, Base, and Arbitrum. The consolidation hub was attacker wallet 1 ([`0xa447f71782135ab96a71374271a749ff7aa54859`](https://etherscan.io/address/0xa447f71782135ab96a71374271a749ff7aa54859)), which consolidated funds via Relay. The funds were bridged, swapped and eventually consolidated as 3,070,767.09 DAI.

At the time of the investigation notes, the DAI had not moved out of attacker wallet.

<br/>

## FAQs

<br/>

> ### Were Axelar or Squid compromised?

Reviewer: Should we have a timeline about the attack including when we were notified by who and when Blockaid tweeted and when Squid responded and when the suspected deployer sent the message onchain?

Author: I can include if you share the data, I have public info like the Blockaid tweet. I didn't think this is super useful but not against including it.

Reviewer: Ok, we can skip

> At the time of the investigation notes, that DAI had not moved out of attacker wallet 5.

Reviewer: ```suggestion
At the time of writing, the consolidated DAI tokens have not been moved out of attacker wallet 5.
```

> ### Were Axelar or Squid compromised?

Reviewer: ```suggestion
### Was Axelar or Squid compromised?
```

> No. The affected component was a third-party Safe application module that incorrectly used Axelar's cross-chain execution protocol. Based on a recent [Blockscan Chat message](https://etherscan.io/tx/0xa7db9793d4978a26ee01c2a6cbf5ba154925a7532e31ae731f7f08ba4295ee70) it's possible that [NewMarketTrading](https://newmarkettrading.com) was the deployer of `SquidRouterModule`.

Reviewer: ```suggestion
No. The affected component was a third-party Safe module called `SquidRouterModule` that incorrectly used Axelar's cross-chain execution protocol. Based on a recent [Blockscan Chat message](https://etherscan.io/tx/0xa7db9793d4978a26ee01c2a6cbf5ba154925a7532e31ae731f7f08ba4295ee70), it's possible that [NewMarketTrading](https://newmarkettrading.com) was the deployer of the affected module.
```

> No. The affected component was a third-party Safe application module that incorrectly used Axelar's cross-chain execution protocol. Based on a recent [Blockscan Chat message](https://etherscan.io/tx/0xa7db9793d4978a26ee01c2a6cbf5ba154925a7532e31ae731f7f08ba4295ee70) it's possible [NewMarketTrading](https://newmarkettrading.com) is the victim platform and the author of these contracts.

Reviewer: ```suggestion
No. The affected component was a third-party Safe application module that incorrectly used Axelar's cross-chain execution protocol. Based on a recent [Blockscan Chat message](https://etherscan.io/tx/0xa7db9793d4978a26ee01c2a6cbf5ba154925a7532e31ae731f7f08ba4295ee70), it's possible that [NewMarketTrading](https://newmarkettrading.com) was the deployer of `SquidRouterModule`.
```

### Did the attacker compromise victim Safe owners?

> Victim owner keys were not compromised. The exploit used the affected Safe module's permissions and an attacker-controlled payload-provided delegate.

Reviewer: ```suggestion
There is no evident to suggest that owner keys were compromised. The exploit used the affected Safe module's permissions and an attacker-controlled payload-provided delegate.
```
